pkg/auth/dependency/interaction/flows/auth_api_sso.go

```diff
@@ -0,0 +1,130 @@
+package flows
+
+import (
+	"errors"
+
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/identity/oauth"
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/interaction"
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/sso"
+	"github.com/skygeario/skygear-server/pkg/core/authn"
+)
+
+func (f *AuthAPIFlow) LoginWithOAuthProvider(oauthAuthInfo sso.AuthInfo, codeChallenge string) (string, *sso.SkygearAuthorizationCode, error) {
+	providerID := oauth.NewProviderID(oauthAuthInfo.ProviderConfig)
+	claims := map[string]interface{}{
+		interaction.IdentityClaimOAuthProvider:  providerID.ClaimsValue(),
+		interaction.IdentityClaimOAuthSubjectID: oauthAuthInfo.ProviderUserInfo.ID,
+		interaction.IdentityClaimOAuthProfile:   oauthAuthInfo.ProviderRawProfile,
+	}
+	i, err := f.Interactions.NewInteractionLogin(&interaction.IntentLogin{
+		Identity: interaction.IdentitySpec{
+			Type:   authn.IdentityTypeOAuth,
+			Claims: claims,
+		},
+	}, "")
+	if errors.Is(err, interaction.ErrInvalidCredentials) {
+		// try signup
+		i, err = f.Interactions.NewInteractionSignup(&interaction.IntentSignup{
+			Identity: interaction.IdentitySpec{
+				Type:   authn.IdentityTypeOAuth,
+				Claims: claims,
+			},
+		}, "")
+		if err != nil {
+			return "", nil, err
+		}
+	} else if err != nil {
+		return "", nil, err
+	}
+
+	s, err := f.Interactions.GetInteractionState(i)
+	if err != nil {
+		return "", nil, err
+	}
+
+	switch s.CurrentStep().Step {
+	case interaction.StepAuthenticateSecondary, interaction.StepSetupSecondaryAuthenticator:
+		panic("interaction_flow_auth_api: TODO: handle MFA")
+	case interaction.StepCommit:
+		attrs, err := f.Interactions.Commit(i)
+		if err != nil {
+			return "", nil, err
+		}
+
+		codeStr, code, err := f.genterateCode(codeChallenge, i, attrs)
```

Comment:
I think we may as well as directly return the interaction token and remove the `SkygearAuthorizationCode` mechanism? We can store the code challenge in interaction state, and defer the commitment of interaction to code exchange.

Reply:
Replacing `SkygearAuthorizationCode` with interaction token is a good idea. But I think we do need commit before code exchange, since user is created in `auth_handler`. 

Do you think it is safe for me to create a new login interaction with `AuthenticatedAs` in `auth_handler`, and use this interaction token as the authorization code?

Btw, I found another problem in my implementation, if user is logging in an existing account  and mfa is required, both `LoginWithOAuthProvider` and `ExchangeCode` will go to `StepAuthenticateSecondary` step.

How about if I change it this way:
In `LoginWithOAuthProvider` which called by `auth_handler`
     - If current step is `StepCommit`, commit it and create new login interaction and return new interaction token as code?
     - If current step is `StepAuthenticateSecondary`, we should return the token. But should we return the original interaction token or create a new login interaction? If we use the original interaction token, we may need resume the interaction in code exchange.

So mfa will be handled in exchange code only

Thought?

Reply:
> Replacing SkygearAuthorizationCode with interaction token is a good idea. But I think we do need commit before code exchange, since user is created in auth_handler.

Yes, sorry for being unclear. I mean the commit of login interaction should be deferred to code exchange. Signup interaction should still be committed in `LoginWithOAuthProvider`.

> Do you think it is safe for me to create a new login interaction with AuthenticatedAs in auth_handler, and use this interaction token as the authorization code?

Yes I think so. That is, `LoginWithOAuthProvider` should always return token for a login interaction, with potential signup interaction already commited.

> If current step is StepCommit, commit it and create new login interaction and return new interaction token as code?

I think at this point, we can just save the interaction and return the token directly.

> If current step is StepAuthenticateSecondary, we should return the token. But should we return the original interaction token or create a new login interaction? If we use the original interaction token, we may need resume the interaction in code exchange.

We should do the same and return token of the login interaction. Simply speaking, `LoginWithOAuthProvider` would just create a new login interaction, save it, and return its token. The interaction should be resumed and processed in code exchange.

Reply:
Update in 4d36753 and ff00aba, see if it looks good:) I will rebase the pr later